examples/pennylane_run_quantum_natural_gradient.py

```diff
@@ -0,0 +1,451 @@
+#
+# .. math::
+#
+#     g_{ij}^{(\ell)} = \langle \psi_{\ell-1} | K_i K_j | \psi_{\ell-1} \rangle
+#     - \langle \psi_{\ell-1} | K_i | \psi_{\ell-1}\rangle
+#     \langle \psi_{\ell-1} |K_j | \psi_{\ell-1}\rangle
+#
+# where
+#
+# .. math::
+#
+#     | \psi_{\ell-1}\rangle = V_{\ell-1}(\theta_{\ell-1}) W_{\ell-1} \cdots V_{0}(\theta_{0}) W_{0} |\psi_0\rangle.
+#
+# (that is, :math:`|\psi_{\ell-1}\rangle` is the quantum state prior to the application
+# of parameterized layer :math:`\ell`), and we have :math:`K_i \equiv K_i^{(\ell)}` for brevity.
+#
+# Let's consider a small variational quantum circuit example coded in PennyLane:
+
+import numpy as np
+
+import pennylane as qml
+from pennylane import expval, var
+
+dev = qml.device('default.qubit', wires=3)
+
+@qml.qnode(dev)
+def circuit(params):
+    # non-parametrized gates
+    qml.RY(np.pi/4, wires=0)
+    qml.RY(np.pi/3, wires=1)
+    qml.RY(np.pi/7, wires=2)
+
+    # Parametrized layer 1
+    qml.RZ(params[0], wires=0)
+    qml.RZ(params[1], wires=1)
+
+    # non-parametrized gates
+    qml.CNOT(wires=[0, 1])
+    qml.CNOT(wires=[1, 2])
+
+    # Parametrized layer 2
+    qml.RY(params[2], wires=1)
+    qml.RX(params[3], wires=2)
+
+    # non-parametrized gates
+    qml.CNOT(wires=[0, 1])
+    qml.CNOT(wires=[1, 2])
+
+    return qml.expval(qml.PauliY(0))
+
+params = np.array([0.432, -0.123, 0.543, 0.233])
+
+##############################################################################
+# The above circuit consists of 4 parameters, with two distinct parametrized
+# layers of 2 parameters each.
+#
+# .. figure:: ../../examples/figures/quantum_natural_gradient/qng1.png
+#     :align: center
+#     :width: 90%
+#     :target: javascript:void(0)
+#
+# |
+#
+# Therefore, the block-diagonal approximation consists of two
+# :math:`2\times 2` matrices, :math:`g^{(1)}` and :math:`g^{(2)}`.
+#
+# .. figure:: ../../examples/figures/quantum_natural_gradient/qng2.png
+#     :align: center
+#     :width: 30%
+#     :target: javascript:void(0)
+#
+# To compute the first block-diagonal :math:`g^{(1)}`, we create subcircuits consisting
+# of all gates prior to the layer, and observables corresponding to
+# the *generators* of the gates in the layer:
+#
+# .. figure:: ../../examples/figures/quantum_natural_gradient/qng3.png
+#     :align: center
+#     :width: 30%
+#     :target: javascript:void(0)
+
+g1 = np.zeros([2, 2])
+
+def layer1_subcircuit(params):
+    """This function contains all gates that
+    precede parametrized layer 1"""
+    qml.RY(np.pi/4, wires=0)
+    qml.RY(np.pi/3, wires=1)
+    qml.RY(np.pi/7, wires=2)
+
+##############################################################################
+# We then post-process the measurement results in order to determine :math:`g^{(1)}`,
+# as follows.
+#
+# .. figure:: ../../examples/figures/quantum_natural_gradient/qng4.png
+#     :align: center
+#     :width: 50%
+#     :target: javascript:void(0)
+#

```

Comment:
Suggestion: put the vertical theta_0 and theta_1 labels on the left-side of matrix (may just be my brain's own internal convention :laughing: )

Reply:
also done lol